ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/assertions/CommandDeduplicationAssertions.scala

```diff
@@ -0,0 +1,118 @@
+
+import java.time.{Duration, Instant}
+
+import com.daml.api.util.DurationConversion
+import com.daml.ledger.api.testtool.infrastructure.Assertions.{assertDefined, fail}
+import com.daml.ledger.api.testtool.infrastructure.participant.{
+  CompletionResponse,
+  ParticipantTestContext,
+}
+import com.daml.ledger.api.v1.completion.Completion
+import com.daml.ledger.api.v1.experimental_features.CommandDeduplicationPeriodSupport.{
+  DurationSupport,
+  OffsetSupport,
+}
+import com.daml.ledger.api.v1.ledger_offset.LedgerOffset
+import com.daml.ledger.client.binding.Primitive.Party
+import com.google.protobuf.duration.{Duration => DurationProto}
+import io.grpc.Status.Code
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits._
+
+object CommandDeduplicationAssertions {
+
+  def assertDeduplicationDuration(
+      requestedDeduplicationDuration: DurationProto,
+      previousSubmissionSendTime: Instant,
+      completionReceiveTime: Instant,
+      completion: Completion,
+      submittingParty: Party,
+      ledger: ParticipantTestContext,
+  )(implicit executionContext: ExecutionContext): Future[Unit] = {
+    val requestedDuration = DurationConversion.fromProto(requestedDeduplicationDuration)
+    ledger.features.commandDeduplicationFeatures.getDeduplicationPeriodSupport.durationSupport match {
+      case DurationSupport.DURATION_NATIVE_SUPPORT =>
+        val reportedDurationProto = assertDefined(
+          completion.deduplicationPeriod.deduplicationDuration,
+          "No deduplication duration has been reported",
+        )
+        val reportedDuration = DurationConversion.fromProto(reportedDurationProto)
+        assert(
+          reportedDuration >= requestedDuration,
+          s"The reported deduplication duration $reportedDuration was smaller than the requested deduplication duration $requestedDuration.",
+        )
+        Future.unit
+      case DurationSupport.DURATION_CONVERT_TO_OFFSET =>
+        val reportedOffset = assertDefined(
+          completion.deduplicationPeriod.deduplicationOffset,
+          "No deduplication offset has been reported",
+        )
+        if (completion.getStatus.code == Code.ALREADY_EXISTS.value) {
+          val completionStreamRequest = ledger.completionStreamRequest(
+            LedgerOffset.of(LedgerOffset.Value.Absolute(reportedOffset))
+          )(submittingParty)
+          ledger
+            .findCompletion(completionStreamRequest)(_.commandId == completion.commandId)
+            .map { optOffsetCompletionResponse =>
+              val _ = assertDefined(
+                optOffsetCompletionResponse,
+                s"No completion with the command ID '${completion.commandId}' after the reported offset $reportedOffset has been found",
+              )
+            }
+        } else {
+          // Let t2 (completionReceiveTime) be the time when the application received the completion offset that specifies the deduplication offset off1.
```

Comment:
What is `off1`? How does `off1` relate to `reportedOffset`?

Reply:
> What is `off1`?

As you can see in the line below, that's an offset for the completed submission (it would be the requested deduplication offset if it was a test for duplication with a deduplication offset, at least while the deduplication offsets are inclusive). That's almost one-to-one from the acceptance criteria.


> How does `off1` relate to reportedOffset?

IMHO it's unrelated, at least while the reported offsets are inclusive.